refactor: validation: mark CheckBlockIndex as const

stickies-v opened the pull request:
While reviewing another PR, I [noticed](https://github.com/bitcoin/bitcoin/pull/31405#discussion_r2056509235) that `ChainstateManager::CheckBlockIndex()`  is not a `const` method. To try and assert that this method was not causing any side-effects, I modified the method to make it `const`. It did not surface any errors, but I think it would be good to merge this change regardless, even if `CheckBlockIndex` is only used in regtest.

This PR removes `CheckBlockIndex()`'s calls to non-const `ChainstateManager` methods by marking `SnapshotBase` `const` and ~inlining the `GetAll()` calls (thereby also performing consistency checks on invalid or fully validated `m_disabled==true` chainstates, as slight behaviour change), and finally marks `CheckBlockIndex()` as `const`.

stickies-v committed:
refactor: validation: mark SnapshotBase as const

It does not modify any logical state. This change is required
for future const-correctness commits.

stickies-v committed:
refactor: validation: mark CheckBlockIndex as const

As a check/test method, this function should not mutate logical
state. Mark it as const to better help ensure this.

mzumsande reviewed (commented):
Concept ACK

Don't really love duplicating `GetAll` in the second commit, but if there is no better solution, marking CBI as const seems more important.

TheCharlatan commented:
Concept ACK

I wrote this simpler approach here https://github.com/TheCharlatan/bitcoin/commit/fe13fa082e3d6befa6d5332428584570ead608ab after seeing you mentioning it last week. I think making sure that all block indexes are declared const in the method might be beneficial though. Maybe instead of adding a separate `GetAll` method just make it const and then declare the chainstate pointers as const in the for loop?

stickies-v commented:
Force-pushed to deduplicate the `GetAll()` overloads by adding a `GetAllImpl()` templated helper, addressing @mzumsande's [comment](https://github.com/bitcoin/bitcoin/pull/32364#pullrequestreview-2804523333).

> I wrote this simpler approach here [TheCharlatan@fe13fa0](https://github.com/TheCharlatan/bitcoin/commit/fe13fa082e3d6befa6d5332428584570ead608ab) after seeing you mentioning it last week. I think making sure that all block indexes are declared const in the method might be beneficial though. Maybe instead of adding a separate `GetAll` method just make it const and then declare the chainstate pointers as const in the for loop?

Allowing a `const ChainstateManager` to modify its `Chainstate` logical state feels like a const-correctness violation to me. If you agree that's true, then I don't think the shorter diff is a worthwhile trade-off. Furthermore, there are other places (such as `GetPruneRange()` that would benefit from a `GetAll() const`, so I think this is worth adding?

shahsb reviewed (commented):
Concept ACK

shahsb commented:
ACK https://github.com/bitcoin/bitcoin/pull/32364/commits/657e58bcd739b52e7b9c9349cd14f05b6b79297c

mzumsande reviewed (commented):
Given that #30214 wants to removes `GetAll()` anyway in 64fc6603525f515c5303f557b038753904c6541(FYI @ryanofsky), the simplest solution might be to not use `GetAll` and just iterate over `m_ibd_chainstate.get()` and `m_snapshot_chainstate.get()` by hand:
```
for (const Chainstate* c : {m_ibd_chainstate.get(), m_snapshot_chainstate.get()}) {
    if (!c) continue;
    (...)
```
this would have the added benefit that we could replace the `IsUsable(cs)` by just a not-nullptr check for cs, because I don't see a conceptual reason why a disabled chainstate (whether it's invalid, or fully validated) shouldn't be subject to `CheckBlockIndex()` anymore.

stickies-v committed:
validation: don't use GetAll() in CheckBlockIndex()

GetAll() is non-const, preventing CheckBlockIndex() from being
const. Rather than add a const GetAll() method, just iterate over
the chainstates directly.

Slight behaviour change by also subjecting non-`IsUsable()`
chainstates to consistency checks.

stickies-v commented:
Force-pushed to remove the `GetAll() const` method (and commit), replacing its callsites with directly accessing `{m_ibd_chainstate.get(), m_snapshot_chainstate.get()}` as [suggested](https://github.com/bitcoin/bitcoin/pull/32364?notification_referrer_id=NT_kwDOBB0EGbQxNjExMTQxNTE5MDo2OTAxMDQ1Nw#pullrequestreview-2810060421) by @mzumsande.

> Given that #30214 wants to removes `GetAll()` anyway in [64fc660](https://github.com/bitcoin/bitcoin/commit/64fc6603525f515c5303f557b038753904c6541e)

Thanks, I hadn't yet reviewed that PR, and the direction seems sensible. I've adopted your suggestion to minimize work in this and conflicting PRs.

mzumsande reviewed (commented):
Code Review ACK 3e6ac5bf772751c66cdcd015dcb7e6ce4ea2cc77

sedited reviewed (approved):
ACK 3e6ac5bf772751c66cdcd015dcb7e6ce4ea2cc77

achow101 commented:
ACK 3e6ac5bf772751c66cdcd015dcb7e6ce4ea2cc77
